pkg/storegateway/series_chunks.go

```diff
@@ -365,24 +387,58 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 
 			err := c.chunkReaders.addLoad(chunk.blockID, chunk.ref, i, j)
 			if err != nil {
-				c.err = errors.Wrap(err, "preloading chunks")
-				return false
+				return nil, false, errors.Wrap(err, "preloading chunks")
 			}
 		}
 	}
 
 	// Create a batched memory pool that can be released all at once.
 	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
 
-	err := c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
+	err = c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
 	if err != nil {
-		c.err = errors.Wrap(err, "loading chunks")
-		return false
+		return nil, false, errors.Wrap(err, "loading chunks")
 	}
 
 	nextSet.chunksReleaser = chunksPool
-	c.current = nextSet
-	return true
+
+	if c.ignoreNativeHistograms {
+		tryNext = filterNativeHistograms(&nextSet)
+		if tryNext {
+			return nil, true, nil
+		}
+	}
+
+	return &nextSet, false, nil
+}
+
+func filterNativeHistograms(nextSet *seriesChunksSet) (tryNext bool) {
```

Comment:
you should be able to make this a function which takes `[]seriesEntry` and returns a `[]seriesEntry`. That way you can reuse it in bucket.go as well; now bucket.go has pretty much the same code

Reply:
Ah, good idea - will do!